cou/utils/openstack.py

```diff
@@ -13,118 +13,125 @@
-        "origin_setting": "openstack-origin",
-    },
-    "ovn-central": {"representative_workload_pkg": "ovn-common", "origin_setting": "source"},
-    "ceph-mon": {"representative_workload_pkg": "ceph-common", "origin_setting": "source"},
-    "placement": {
-        "representative_workload_pkg": "placement-common",
-        "origin_setting": "openstack-origin",
-    },
-    "nova-cloud-controller": {
-        "representative_workload_pkg": "nova-common",
-        "origin_setting": "openstack-origin",
-    },
-}
-
-
-def get_os_code_info(package, pkg_version) -> str:
-    """Determine OpenStack codename that corresponds to package version.
-
-    :param package: Package name
-    :type package: string
-    :param pkg_version: Package version
-    :type pkg_version: string
-    :returns: Codename for package
-    :rtype: string
-    """
-    # Remove epoch if it exists
-    if ":" in pkg_version:
-        pkg_version = pkg_version.split(":")[1:][0]
-    if "swift" in package:
-        # Fully x.y.z match for swift versions
-        match = re.match(r"^(\d+)\.(\d+)\.(\d+)", pkg_version)
-    else:
-        # x.y match only for 20XX.X
-        # and ignore patch level for other packages
-        match = re.match(r"^(\d+)\.(\d+)", pkg_version)
-
-    if match:
-        vers = match.group(0)
-    # Generate a major version number for newer semantic
-    # versions of openstack projects
-    major_vers = vers.split(".")[0]
-    if package in PACKAGE_CODENAMES and major_vers in PACKAGE_CODENAMES[package]:
-        return PACKAGE_CODENAMES[package][major_vers]
-    else:
-        # < Liberty co-ordinated project versions
-        if "swift" in package:
-            return get_swift_codename(vers)
-        elif "ovn" in package:
-            return get_ovn_codename(vers)
-        else:
-            return OPENSTACK_CODENAMES[vers]
+"""Lookup utils to have the latest compatible Openstack codename based on workload version."""
 
+import csv
+import logging
+from collections import OrderedDict, defaultdict
+from dataclasses import dataclass
+from typing import Any, Union
 
-# Codename and package versions
-def get_swift_codename(version):
-    """Determine OpenStack codename that corresponds to swift version.
+from packaging.version import Version
 
-    :param version: Version of Swift
-    :type version: string
-    :returns: Codename for swift
-    :rtype: string
-    """
-    return _get_special_codename(version, SWIFT_CODENAMES)
 
+@dataclass(frozen=True)
+class Versions:
+    min_version: Version
+    max_version: Version
+    minor: bool = False
+    micro: bool = False
 
-def get_ovn_codename(version):
-    """Determine OpenStack codename that corresponds to OVN version.
 
-    :param version: Version of OVN
-    :type version: string
-    :returns: Codename for OVN
-    :rtype: string
-    """
-    return _get_special_codename(version, OVN_CODENAMES)
+def generate_openstack_lookup() -> OrderedDict:
```

Comment:
I agree. On the other hand, as a user of the library it would be a lot more natural to import and use a function directly (e.g. `get_compatible_codename` instead of having to first instantiate a "Comparer" object. We could still handle `generate_openstack_lookup` as a class `__init__` internally though

Reply:
The `generate_openstack_lookup` could return the object itself, however I agree that with Gabriel that we do not need to initialized it every-time when someone call `generate_openstack_lookup`, so we can follow singleton pattern here.